python/lsst/sims/utils/ZernikeModule.py

```diff
@@ -0,0 +1,243 @@
+
+        n_coeffs = 1+(n-m)//2
+        local_coeffs = np.zeros(n_coeffs, dtype=float)
+        local_powers = np.zeros(n_coeffs, dtype=float)
+        for k in range(0, n_coeffs):
+            if k%2 == 0:
+                sgn = 1.0
+            else:
+                sgn = -1.0
+
+            num_fac = self._factorial.evaluate(n-k)
+            k_fac = self._factorial.evaluate(k)
+            d1_fac = self._factorial.evaluate(((n+m)//2)-k)
+            d2_fac = self._factorial.evaluate(((n-m)//2)-k)
+
+            local_coeffs[k] = sgn*num_fac/(k_fac*d1_fac*d2_fac)
+            local_powers[k] = n-2*k
+
+        self._coeffs[(n,m)] = local_coeffs
+        self._powers[(n,m)] = local_powers
+
+    def _evaluate_radial_number(self, r, nm_tuple):
+        """
+        Evaluate the radial part of a Zernike polynomial.
+
+        r is a scalar value
+
+        nm_tuple is a tuple of the form (radial order, angular order)
+        denoting the polynomial to evaluate
+
+        Return the value of the radial part of the polynomial at r
+        """
+        r_term = np.power(r, self._powers[nm_tuple])
+        return (self._coeffs[nm_tuple]*r_term).sum()
+
+    def _evaluate_radial_array(self, r, nm_tuple):
+        """
+        Evaluate the radial part of a Zernike polynomial.
+
+        r is a numpy array of radial values
+
+        nm_tuple is a tuple of the form (radial order, angular order)
+        denoting the polynomial to evaluate
+
+        Return the values of the radial part of the polynomial at r
+        """
+        # since we use np.where to handle cases of
+        # r==0, use np.errstate to temporarily
+        # turn off the divide by zero and
+        # invalid double scalar RuntimeWarnings
+        with np.errstate(divide='ignore', invalid='ignore'):
+            log_r = np.log(r)
+            log_r = np.where(np.isfinite(log_r), log_r, -1.0e10)
+            r_power = np.exp(np.outer(log_r, self._powers[nm_tuple]))
+        return np.dot(r_power, self._coeffs[nm_tuple])
+
+    def _evaluate_radial(self, r, n, m):
+        """
+        Evaluate the radial part of a Zernike polynomial
+
+        r is a radial value or an array of radial values
+
+        n is the radial order of the polynomial
+
+        m is the angular order of the polynomial
+
+        Return the value(s) of the radial part of the polynomial at r
+        """
+
+        is_array = False
+        if not isinstance(r, numbers.Number):
+            is_array = True
+
+        nm_tuple = self._validate_nm(n,m)
+
+        if (nm_tuple[0]-nm_tuple[1]) % 2 == 1:
+            if is_array:
+                return np.zeros(len(r), dtype=float)
+            return 0.0
+
+        if nm_tuple not in self._coeffs:
+            self._make_polynomial(nm_tuple[0], nm_tuple[1])
+
+        if is_array:
+            return self._evaluate_radial_array(r, nm_tuple)
+
+        return self._evaluate_radial_number(r, nm_tuple)
+
+    def evaluate(self, r, phi, n, m):
```

Comment:
Just following from the wiki page where 0 <= r <= 1, should we have a check for r<=1.0? Is that implicitly in here somewhere and I missed it?

Reply:
I was not enforcing r<=1.0;  I have added that